File: sources/img/wavelet_image.cc
// Copyright 2020-2022 PANDA GmbH

#include "wavelet_buffer/img/wavelet_image.h"

#include <fstream>
#include <utility>
#include <vector>

#include "wavelet_buffer/wavelet_buffer_view.h"
#include "wavelet_buffer/wavelet_utils.h"

namespace drift::img {

class WaveletImage::WaveletImageImpl {
 public:
  explicit WaveletImageImpl(WaveletParameters parameters)
      : buffer_(std::move(parameters)) {}

  explicit WaveletImageImpl(WaveletImage::Buffer buffer)
      : buffer_(std::move(buffer)) {}

  WaveletImage::Status ImportFromFile(const std::string &file_path,
                                      const Denoiser &denoiser,
                                      const Codec &codec,
                                      size_t start_channel) {
    Status status;

    std::ifstream file(file_path, std::ios::binary);
    if (file) {
      std::stringstream ss;
      ss << file.rdbuf();
      status = ImportFromString(ss.str(), denoiser, codec, start_channel);
    } else {
      status = Status::kIOError;
    }

    return status;
  }

  Status ExportToFile(const std::string &file_path, const Codec &codec,
                      size_t start_channel) const {
    Status status;

    std::ofstream file(file_path, std::ios::binary);
    if (file) {
      std::string data;
      status = ExportToString(&data, codec, start_channel);
      if (status == Status::kOk) {
        file << data;
      }
    } else {
      status = Status::kIOError;
    }

    return status;
  }

  WaveletImage::Status ImportFromString(const std::string &data,
                                        const Denoiser &denoiser,
                                        const Codec &codec,
                                        size_t start_channel) {
    Status status = Status::kOk;

    SignalN2D image;
    if (codec.Decode(data, &image)) {
      if (!buffer_(start_channel, image.size()).Decompose(image, denoiser)) {
        status = Status::kDecompositionError;
      }
    } else {
      status = Status::kWrongData;
    }

    return status;
  }

  WaveletImage::Status ExportToString(std::string *data, const Codec &codec,
                                      size_t start_channel) const {
    auto status = Status::kOk;
    SignalN2D image;

    if (buffer_(start_channel, codec.channel_number()).Compose(&image)) {
      if (!codec.Encode(image, data)) {
        status = Status::kWrongData;
      }
    } else {
      status = Status::kCompositionError;
    }

    return status;
  }

  Status Save(const std::string &file_path) const {
    Status status;
    std::ofstream file(file_path, std::ios::binary);
    if (file) {
      std::string data;
      status = buffer_.Serialize(&data) ? Status::kOk : Status::kWrongData;
      if (status == Status::kOk) {
        file << data;
      }
    } else {
      status = Status::kIOError;
    }

    return status;
  }

  Buffer buffer_;
};

WaveletImage::WaveletImage(size_t channel_number, size_t w, size_t h)
    : impl_(std::make_unique<WaveletImageImpl>(WaveletParameters{
          .signal_shape = {w, h},
          .signal_number = channel_number,
          .decomposition_steps = 0,
          .wavelet_type = WaveletTypes::kDB1,
      })) {}

WaveletImage::WaveletImage(WaveletParameters parameters)
    : impl_(std::make_unique<WaveletImageImpl>(std::move(parameters))) {}

WaveletImage::WaveletImage(WaveletImage::Buffer buffer)
    : impl_(std::make_unique<WaveletImageImpl>(std::move(buffer))) {}

WaveletImage::~WaveletImage() = default;

WaveletImage::Status WaveletImage::ImportFromFile(const std::string &file_path,
                                                  const Denoiser &denoiser,
                                                  const Codec &codec,
                                                  size_t start_channel) {
  return impl_->ImportFromFile(file_path, denoiser, codec, start_channel);
}

WaveletImage::Status WaveletImage::ExportToFile(const std::string &file_path,
                                                const Codec &codec,
                                                size_t start_channel) const {
  return impl_->ExportToFile(file_path, codec, start_channel);
}

WaveletImage::Status WaveletImage::ImportFromString(const std::string &data,
                                                    const Denoiser &denoiser,
                                                    const Codec &codec,
                                                    size_t start_channel) {
  return impl_->ImportFromString(data, denoiser, codec, start_channel);
}

WaveletImage::Status WaveletImage::ExportToString(std::string *data,
                                                  const Codec &codec,
                                                  size_t start_channel) const {
  return impl_->ExportToString(data, codec, start_channel);
}

std::unique_ptr<WaveletImage> WaveletImage::Load(const std::string &file_path) {
  std::unique_ptr<WaveletImage> img;

  std::ifstream file(file_path, std::ios::binary);
  if (file) {
    std::stringstream ss;
    ss << file.rdbuf();
    auto buffer = Buffer::Parse((ss.str()));
    if (buffer) {
      img = std::make_unique<WaveletImage>(std::move(*buffer));
    }
  }

  return img;
}

WaveletImage::Status WaveletImage::Save(const std::string &file_path) const {
  return impl_->Save(file_path);
}

const WaveletImage::Buffer &WaveletImage::const_buffer() const {
  return impl_->buffer_;
}

WaveletImage::Buffer &WaveletImage::buffer() { return impl_->buffer_; }

double WaveletImage::CompareTo(const WaveletImage &other) const {
  return Distance(const_buffer(), other.const_buffer());
}

}  // namespace drift::img

File: python/src/wavelet_buffer/img/wavelet_image.py
"""Helper module to use WaveletBuffer with images"""
from typing import List
import wavelet_buffer._wavelet_buffer as impl

from wavelet_buffer.denoise import BaseDenoiser
from wavelet_buffer.img.codecs import BaseCodec
from wavelet_buffer.wavelet_buffer import WaveletType, WaveletBuffer


class WaveletImage(impl.WaveletImage):
    """A class to load/save images of different formats and store them in WaveletBuffer"""

    def __init__(
        self,
        signal_shape: List[int],
        signal_number: int,
        decomposition_steps: int,
        wavelet_type: WaveletType,
    ):
        """
        Args:
            signal_shape: for 1D signal with size N [N], for 2DN signal with size MXN, [M,N]
            signal_number: N for 2DN signal, e.g. RGB image would be 3 signals
            decomposition_steps: number of decomposition steps
            wavelet_type: type WaveletType.NONE, DB{1..5}, if it is NONE no wavelet composition
        """
        super().__init__(signal_shape, signal_number, decomposition_steps, wavelet_type)

    def import_from_file(
        self,
        file_path: str,
        denoiser: BaseDenoiser,
        codec: BaseCodec,
        start_channel: int = 0,
    ) -> None:
        """Import an image from file and decompose it with WaveletBuffer
        Args:
            file_path: path to image
            denoiser: denoising algorithm
            codec: a coodec for decoding the image format
            start_channel: need if you want to decode specific channel as a monochrome image
        """
        super(WaveletImage, self).import_from_file(
            file_path, denoiser, codec._impl, start_channel
        )

    def export_to_file(
        self, file_path: str, codec: BaseCodec, start_channel: int = 0
    ) -> None:
        """Encoding a wavelet image to a file with common image format e.g. JPEG
        Args:
            file_path: path to image
            codec:  a coodec for encoding the image format
            start_channel: need if you want to encode specific channel as a monochrome image, or you have
             a few images in the buffer
        """
        super(WaveletImage, self).export_to_file(file_path, codec._impl, start_channel)

    def import_from_string(
        self,
        data: bytes,
        denoiser: BaseDenoiser,
        codec: BaseCodec,
        start_channel: int = 0,
    ) -> None:
        """Import an image from byte sting and decompose it with WaveletBuffer
        Args:
            data: encoded image e.g. JPEG string
            denoiser: denoising algorithm
            codec: a coodec for decoding the image format
            start_channel: need if you want to decode specific channel as a monochrome image
        """
        super(WaveletImage, self).import_from_string(
            data, denoiser, codec._impl, start_channel
        )

    def export_to_string(self, codec: BaseCodec, start_channel: int = 0) -> bytes:
        """Encoding a wavelet image to a byte sting with common image format e.g. JPEG
        Args:
            codec:  a coodec for encoding the image format
            start_channel: need if you want to encode specific channel as a monochrome image, or you have
             a few images in the buffer
        Returns:
            encoded image e.g. JPEG string
        """
        return super(WaveletImage, self).export_to_string(codec._impl, start_channel)

    @staticmethod
    def load(file_path: str) -> "WaveletImage":
        """Deserialize a wavelet image from file
        Args:
            file_path:
        Returns:
            decompressed image
        """
        return impl.WaveletImage.load(file_path)

    def save(self, file_path: str) -> None:
        """Serialize image to a file
        Args:
            file_path:
        """
        super(WaveletImage, self).save(file_path)

    def distance(self, other: "WaveletImage") -> float:
        """Distance to other image. 1 - the same, 0 - something completely different"""
        return super(WaveletImage, self).distance(other)

    @property
    def buffer(self) -> WaveletBuffer:
        """Get wavelet buffer"""
        return super(WaveletImage, self).buffer
